### ISSUE TITLE
Build regressions on 25.3.26 (vs 25.3.26) on hppa, hurd-i386, kfreebsd-*

### ISSUE THREAD

**Issue by ottok** [codership/galera#539]
Official Debian builder overview of 25.3.25:
![image](https://user-images.githubusercontent.com/668724/56636115-7235a780-6670-11e9-99d1-eaa715e1a60c.png)
==> All but old powerpc and sparc64 built OK

Status for 25.3.26:
![image](https://user-images.githubusercontent.com/668724/56636146-88dbfe80-6670-11e9-86af-7670ad56e3cc.png)
==> Regressions on hppa, hurd-i386, kfreebsd-*

The hurd and kfreebsd had a bunch of patches made for 25.3.25:
- #525 
- #527 
- #528 

They are still applied in 25.3.26 (https://salsa.debian.org/mariadb-team/galera-3/tree/master/debian/patches) but do not seem to be enough anymore.. Also they have not been merged here upstream.

The mentioned architectures are not widely used, but often build issues on them are caused by non-standard C/C++ code usage, and fixing build failures on them can improve the C/C++ code quality in general and make build or security issues on regular amd64 platforms less likely.

For more information please see the latest build logs themselves at https://buildd.debian.org/status/package.php?p=galera-3 and downstream issue https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=927846

**Comment by temeo** [comment]
Hi Otto!

I scanned the latest builds from https://buildd.debian.org/status/package.php?p=galera-3. The hppa seems to be now passing. 

Regressing builds are:

hurd-i386:
```
Running suite(s): Defaults
failed to open gcomm backend connection: 1073741902: Failed to set thread schedparams other:0: 1073741902 (Function not implemented)
	 at galerautils/src/gu_thread.cpp:thread_set_schedparam():91
```
kfreebsd-amd64:
```
Running suite(s): Defaults
failed to open gcomm backend connection: 78: Failed to set thread schedparams other:0: 78 (Function not implemented)
	 at galerautils/src/gu_thread.cpp:thread_set_schedparam():91
```
kfreebsd-i386:
```
Running suite(s): Defaults
failed to open gcomm backend connection: 78: Failed to set thread schedparams other:0: 78 (Function not implemented)
	 at galerautils/src/gu_thread.cpp:thread_set_schedparam():91
```

These are quite likely caused by new test for default parameters, which cause calls to system calls which are not supported on the platform. I guess we either need to disable schedparam functionality on those platforms or figure out how to handle errors gracefully in the tests.

**Comment by temeo** [comment]
Hi @ottok !

Could you check if the following test fixes the following patch fixes the regression in hurd-i386, kfreebsd-amd64, kfreebsd-i386 builds?

```
diff --git a/galerautils/src/gu_thread.cpp b/galerautils/src/gu_thread.cpp
index e307f6e..40a956e 100644
--- a/galerautils/src/gu_thread.cpp
+++ b/galerautils/src/gu_thread.cpp
@@ -7,6 +7,7 @@
 #include "gu_utils.hpp"
 #include "gu_string_utils.hpp"
 #include "gu_throw.hpp"
+#include "gu_logger.hpp"
 
 #include <iostream>
 #include <vector>
@@ -78,8 +79,11 @@ gu::ThreadSchedparam gu::thread_get_schedparam(pthread_t thd)
     return ThreadSchedparam(policy, sp.sched_priority);
 }
 
+static bool schedparam_not_supported(false);
+
 void gu::thread_set_schedparam(pthread_t thd, const gu::ThreadSchedparam& sp)
 {
+    if (schedparam_not_supported) return;
 #if defined(__sun__)
     struct sched_param spstr = { sp.prio(), { 0, } /* sched_pad array */};
 #else
@@ -88,6 +92,17 @@ void gu::thread_set_schedparam(pthread_t thd, const gu::ThreadSchedparam& sp)
     int err;
     if ((err = pthread_setschedparam(thd, sp.policy(), &spstr)) != 0)
     {
-        gu_throw_error(err) << "Failed to set thread schedparams " << sp;
+        if (err == ENOSYS)
+        {
+            log_warn << "Function pthread_setschedparam() is not implmented "
+                     << "in this system. Future attempts to change scheduling "
+                     << "priority will be no-op";
+
+            schedparam_not_supported = true;
+        }
+        else
+        {
+            gu_throw_error(err) << "Failed to set thread schedparams " << sp;
+        }
     }
 }
```

**Comment by ottok** [comment]
Sorry for the long delay in replying. I uploaded a version with [this patch](https://salsa.debian.org/mariadb-team/galera-3/commit/748b6efbb1c952d0fba0f7b7bef01b924ef40e1c) included to Debian experimental, and the build logs are visible at https://buildd.debian.org/status/package.php?p=galera-3&suite=experimental

I have not removed any of the The hurd and kfreebsd had a bunch of patches made for 25.3.25 (#525, #527, #528).

Hurd-i386 seems to fail now with:
```
failed to open gcomm backend connection: 1073741902: Failed to set thread schedparams other:0: 1073741902 (Function not implemented)
	 at galerautils/src/gu_thread.cpp:thread_set_schedparam():91
gcs/src/gcs_core.cpp:gcs_core_open():209: Failed to open backend connection: -1073741902 (Function not implemented)
gcs/src/gcs.cpp:gcs_open():1458: Failed to open channel 'cluster_name' at 'gcomm://': -1073741902 (Function not implemented)
gcs connect failed: Function not implemented
```

Hppa failed now with:
```
83%: Checks: 6, Failures: 0, Errors: 1
galerautils/tests/gu_rset_test.cpp:355:E:RecordSet v1:ver1:0: (after this point) Test timeout expired
```

In addition there is:
```
gcc -o .sconf_temp/conftest_2.o -c -std=c99 -fno-strict-aliasing -pipe -g -O2 -fdebug-prefix-map=/<<PKGBUILDDIR>>=. -fstack-protector-strong -Wformat -Werror=format-security -g -O3 -DNDEBUG -m32 -march=i686 -pthread -fPIC -Wall -Wextra -Wno-unused-parameter -Wdate-time -D_FORTIFY_SOURCE=2 -D_XOPEN_SOURCE=600 -DHAVE_COMMON_H -Iwsrep/src .sconf_temp/conftest_2.c
.sconf_temp/conftest_2.c:2:10: fatal error: sys/epoll.h: No such file or directory
 #include <sys/epoll.h>
          ^~~~~~~~~~~~~
```

Also, both hurd and hppa had:
```
g++ -o .sconf_temp/conftest_6.o -c -pipe -Wno-long-long -Wno-deprecated -ansi -g -O2 -fdebug-prefix-map=/<<PKGBUILDDIR>>=. -fstack-protector-strong -Wformat -Werror=format-security -g -O3 -DNDEBUG -m32 -march=i686 -pthread -fPIC -Wall -Wextra -Wno-unused-parameter -Wdate-time -D_FORTIFY_SOURCE=2 -D_XOPEN_SOURCE=600 -DHAVE_COMMON_H -DHAVE_BYTESWAP_H -DHAVE_ENDIAN_H -DHAVE_EXECINFO_H -Iwsrep/src .sconf_temp/conftest_6.cpp
.sconf_temp/conftest_6.cpp:3:2: error: #error Not compiling in C++11 mode
 #error Not compiling in C++11 mode
  ^~~~~
```

**Comment by ottok** [comment]
After waiting for 6 days, the kfreebsd builds are still pending. I will stop tracking this for now.
![image](https://user-images.githubusercontent.com/668724/59442560-c0f2f880-8dfa-11e9-90d8-667229e9b266.png)

https://buildd.debian.org/status/package.php?p=galera-3&suite=experimental

**Comment by ottok** [comment]
Success! I can confirm the patches fixed all 3 platforms:
![image](https://user-images.githubusercontent.com/668724/61954134-8d28f800-af8e-11e9-8177-925c26f60aa1.png)


However ia64 regressed... Might be related or not.
```
Running suite(s): saved_state
33%: Checks: 3, Failures: 1, Errors: 1
galera/tests/saved_state_check.cpp:48:E:saved_state:test_unsafe:0: (after this point) Test timeout expired
galera/tests/saved_state_check.cpp:163:F:saved_state:test_corrupt:0: Failure 'uuid == WSREP_UUID_UNDEFINED' occurred
Running suite(s): Defaults
It may not be safe to bootstrap the cluster from this node. It was not the last one to leave the cluster and may not contain all the updates. To force cluster bootstrap with this node, edit the grastate.dat file manually and set safe_to_bootstrap to 1 .
0%: Checks: 1, Failures: 1, Errors: 0
galera/tests/defaults_check.cpp:262:F:defaults:defaults:0: connect() returned 7
Total tests failed: 3
scons: *** [galera/tests/galera_check.passed] Error 1
scons: building terminated because of errors.
...
scons: Configure: Checking if compiling in C++11 mode ... 
.sconf_temp/conftest_7.cpp <-
  |
  |#if __cplusplus < 201103
  |#error Not compiling in C++11 mode
  |#endif
  |int main() { return 0; }
  |
g++ -o .sconf_temp/conftest_7.o -c -pipe -Wno-long-long -Wno-deprecated -ansi -g -O2 -fdebug-prefix-map=/<<PKGBUILDDIR>>=. -Wformat -Werror=format-security -g -O3 -DNDEBUG -pthread -fPIC -Wall -Wextra -Wno-unused-parameter -Wdate-time -D_FORTIFY_SOURCE=2 -D_XOPEN_SOURCE=600 -DHAVE_COMMON_H -DGALERA_USE_GU_NETWORK -DHAVE_BYTESWAP_H -DHAVE_ENDIAN_H -DHAVE_EXECINFO_H -Iwsrep/src .sconf_temp/conftest_7.cpp
.sconf_temp/conftest_7.cpp:3:2: error: #error Not compiling in C++11 mode
 #error Not compiling in C++11 mode
  ^~~~~
....
scons: *** Error 1
scons: *** [libgalera_smm.so] Error 1
scons: building terminated because of errors.
```
https://buildd.debian.org/status/fetch.php?pkg=galera-3&arch=ia64&ver=25.3.27-1%7Eexp1&stamp=1564091840&raw=0

As usual, full details at https://buildd.debian.org/status/package.php?p=galera-3&suite=experimental

**Comment by temeo** [comment]
The fix will be included in the upcoming 3.28.

**Comment by ottok** [comment]
Uploaded 25.3.28 to Debian, build status:
Before (.27):
![image](https://user-images.githubusercontent.com/668724/65013989-d6118300-d925-11e9-9030-42c0e42a894a.png)

Now: (.28):
![image](https://user-images.githubusercontent.com/668724/65014002-df9aeb00-d925-11e9-9023-cc68f4e3ef62.png)

See https://buildd.debian.org/status/package.php?p=galera-3
